language/move-binary-format/src/file_format.rs

```diff
@@ -1935,6 +1938,22 @@ impl CompiledModule {
     }
 }
 
+impl Into<Vec<u8>> for &CompiledModule {
+    fn into(self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        self.serialize(&mut bytes)
+            .expect("serializing well-formed module should never fail");
```

Comment:
This is the dark side of trying to implement `Hash` manually--serialization is not guaranteed to succeed. Which leaves us with two choices:
- Unwrap (as implemented here). Since serialization should only fail on modules that have been created by an adversary or produced by a buggy compiler (i.e., for a published module, serialization should always succeed), this probably ok. Basically, you'll only hit the unwrap if you're doing something wrong.
- Implement `TryInto` and a non-standard variant of `Hash` that returns an `Option`. This would put me back to the drawing board for my use-case (since I need to embed a `CompiledModule` inside a different struct that's deriving `Hash`), but I could hack around it by implementing a wrapper for `CompiledModule` that implements `Into` and `Hash` in the same way as here.

Reply:
Could you just store the CompiledModule bytes? And then hash those? That is, serialize the compiled module and then store those bytes directly and hash those?

Reply:
I could, and that's what I'm doing now. It's just less type-safe and less convenient. E.g., my code looks something like:

```
#[derive(Hash, ...)]
enum Transaction {
  Publish { module: Vec<u8>, ... }
  ...
```

, whereas I'd like it to look like
```
#[derive(Hash, ...)]
enum Transaction {
  Publish { module: CompiledModule, ... }
  ...
```

to save both the clients and validator code the effort of converting between bytes and modules.